### PR TITLE
[feat:] Add Log tailing

### PR DESCRIPTION
**Pull request by srikary12**
### Description
Adds log tailing capability

### Linked issues
https://github.com/jozu-ai/kitops/issues/591

### CODE REVIEW

File: pkg/lib/harness/llm-harness.go [hunk 1]
```diff
@@ -181,8 +183,26 @@ func PrintLogs(configHome string, w io.Writer) error {
 		return fmt.Errorf("error reading log file: %w", err)
 	}
 	defer logFile.Close()
-	if _, err = io.Copy(w, logFile); err != nil {
-		return fmt.Errorf("failed to print log file: %w", err)
+	if !tail {
+		if _, err = io.Copy(w, logFile); err != nil {
+			return fmt.Errorf("failed to print log file: %w", err)
+		}
```
Comment (gorkem) [inline pkg/lib/harness/llm-harness.go:189]: You can simplify early returning in here.  Alternatively you can handle the `!tail` case in the loop.
```
return nil
```

File: pkg/lib/harness/llm-harness.go [hunk 2]
```diff
@@ -181,8 +183,26 @@ func PrintLogs(configHome string, w io.Writer) error {
...
+			if err != nil {
+				if err == io.EOF {
+					time.Sleep(1 * time.Second)
+				} else {
+					return fmt.Errorf("failed to print log file: %w", err)
+				}
+			}
+			stringReader := strings.NewReader(line)
+			if _, err = io.Copy(w, stringReader); err != nil {
+				return fmt.Errorf("failed to print log file: %w", err)
```
Comment (gorkem) [inline pkg/lib/harness/llm-harness.go:203]: This is unncessary overhead you can write the line here instead
```suggestion
	        if _, err := w.Write([]byte(line)); err != nil {
                return fmt.Errorf("failed to write to output: %w", err) 
```

**Review by gorkem — CHANGES_REQUESTED** [review]
This is not handling the case when the server is stopped and started. I expect the log following to either drop or switch to the newly started server instance.

**Comment by srikary12** [comment]
I can drop the logs when server is stopped.

Do we cuurently allow 2 development servers running at once?

**Comment by gorkem** [comment]
> I can drop the logs when server is stopped.
> 
> Do we currently allow 2 development servers running at once?

Only one instance is allowed to run.

**Comment by srikary12** [comment]
@gorkem shows a server stopped message when the server is stopped and resolves conversations :).

File: pkg/lib/harness/llm-harness.go [hunk 3]
```diff
@@ -181,10 +183,35 @@ func PrintLogs(configHome string, w io.Writer) error {
...
+			}
+		}
+		if _, err := w.Write([]byte(line)); err != nil {
+			return fmt.Errorf("failed to write to output: %w", err)
+		}
 	}
-	return nil
+}
+
+func checkExistance(configHome string) bool {
```
Comment (gorkem) [inline pkg/lib/harness/llm-harness.go:210]: Can you fix the spelling? It should be `checkExistence`

**Review by gorkem — CHANGES_REQUESTED** [review]
Realized my inline comments are not uploaded to GH trying again

File: pkg/lib/harness/llm-harness.go [hunk 4]
```diff
@@ -181,10 +183,35 @@ func PrintLogs(configHome string, w io.Writer) error {
...
 	}
-	return nil
+}
+
+func checkExistence(configHome string) bool {
+	pidFile := filepath.Join(constants.HarnessPath(configHome), constants.HarnessProcessFile)
+
+	_, err := readPIDFromFile(pidFile)
+	return os.IsNotExist(err)
 }
```
Comment (amisevsk) [inline pkg/lib/harness/llm-harness.go:215]: This is just checking if the PID file exists, not whether the server is actively running. I think we also need to use `isProcessRunning(pid)`

Comment (srikary12) [reply pkg/lib/harness/llm-harness.go:215]: When we run `kit dev stop` we are verifying the the running server this way, and stopping it. Used the same logic in here.

Comment (srikary12) [reply pkg/lib/harness/llm-harness.go:215]: `func (harness *LLMHarness) Stop() error {
	pidFile := filepath.Join(constants.HarnessPath(harness.ConfigHome), constants.HarnessProcessFile)

	pid, err := readPIDFromFile(pidFile)
	if os.IsNotExist(err) {
		return fmt.Errorf("no Running server found")
	}
	if err != nil {
		return err
	}
`

This is how we are checking the server status in stop interface.

File: pkg/lib/harness/llm-harness.go [hunk 5]
```diff
@@ -181,10 +183,35 @@ func PrintLogs(configHome string, w io.Writer) error {
...
+			if err == io.EOF {
+				if !follow {
+					return nil
+				}
+				time.Sleep(1 * time.Second)
+				serverStatus := checkExistence(configHome)
+				if serverStatus {
+					return fmt.Errorf("server stopped")
+				}
+				continue
```
[the remark this answers is not in the document]
Comment (srikary12) [reply pkg/lib/harness/llm-harness.go:199]: Will resolve this

**Comment by srikary12** [comment]
@amisevsk We still need to show logs even if server crashes. I think including the above case wouldn't display logs when the server crashes. We only remove pid file once server ends properly until then we display logs.

File: pkg/lib/harness/llm-harness.go (hunk 4 above)
Comment (srikary12) [reply pkg/lib/harness/llm-harness.go:215]: We still need to show logs even if server crashes. I think including the above case wouldn't display logs when the server crashes. We only remove pid file once server ends properly until then we display logs.

Comment (amisevsk) [reply pkg/lib/harness/llm-harness.go:215]: These checks run only when reaching EOF on the file though -- we've already printed all existing logs at this point. If the server has crashed, there aren't going to be any more logs to print (same as the server terminated normally case)

Comment (srikary12) [reply pkg/lib/harness/llm-harness.go:215]: Okay, will add that case too.

**Pull request merged**